// musicbrainz-api/lib/coverartarchive-api.js
/* eslint-disable-next-line */
import got from 'got';
export class CoverArtArchiveApi {
    constructor() {
        this.host = 'coverartarchive.org';
    }
    async getJson(path) {
        const response = await got.get('https://' + this.host + path, {
            headers: {
                Accept: `application/json`
            },
            responseType: 'json'
        });
        return response.body;
    }
    /**
     * Fetch release
     * @releaseId Release MBID
     * @param releaseId MusicBrainz Release MBID
     * @param coverType Cover type
     */
    getReleaseCovers(releaseId, coverType) {
        return this.getCovers(releaseId, 'release', coverType);
    }
    /**
     * Fetch release-group
     * @releaseGroupId Release-group MBID
     * @param releaseGroupId MusicBrainz Release Group MBID
     * @param coverType Cover type
     */
    getReleaseGroupCovers(releaseGroupId, coverType) {
        return this.getCovers(releaseGroupId, 'release-group', coverType);
    }
    /**
     * Fetch covers
     * @releaseId MBID
     * @param releaseId MusicBrainz Release Group MBID
     * @param releaseType Fetch covers for specific release or release-group
     * @param coverType Cover type
     */
    async getCovers(releaseId, releaseType = 'release', coverType) {
        const path = [releaseType, releaseId];
        if (coverType) {
            path.push(coverType);
        }
        const info = await this.getJson('/' + path.join('/'));
        // Hack to correct http addresses into https
        if (info.release && info.release.startsWith('http:')) {
            info.release = 'https' + info.release.substring(4);
        }
        return info;
    }
}
